heartbeat/sender/multikeyPeerAuthenticationSender.go

```diff
@@ -0,0 +1,307 @@
+	pkBytes := []byte(pk)
+	if !sender.processIfShouldSend(pkBytes, currentTimeAsUnix) {
+		return nil
+	}
+
+	currentTimeStamp := time.Unix(currentTimeAsUnix, 0)
+
+	data, isHardforkTriggered, err := sender.prepareMessage([]byte(pk), sk)
+	if err != nil {
+		setTimeErr := sender.keysHolder.SetNextPeerAuthenticationTime(pkBytes, currentTimeStamp.Add(sender.timeBetweenSendsWhenError))
+		if setTimeErr != nil {
+			return fmt.Errorf("%w while seting next peer authentication time, after prepare message error %s", setTimeErr, err.Error())
+		}
+		return err
+	}
+	if isHardforkTriggered {
+		nextTimeStamp := currentTimeStamp.Add(sender.computeRandomDuration(sender.hardforkTimeBetweenSends))
+		setTimeErr := sender.keysHolder.SetNextPeerAuthenticationTime(pkBytes, nextTimeStamp)
+		if setTimeErr != nil {
+			return fmt.Errorf("%w while seting next peer authentication time, hardfork triggered", setTimeErr)
+		}
+	} else {
+		nextTimeStamp := currentTimeStamp.Add(sender.computeRandomDuration(sender.timeBetweenSends))
+		setTimeErr := sender.keysHolder.SetNextPeerAuthenticationTime(pkBytes, nextTimeStamp)
+		if setTimeErr != nil {
+			return fmt.Errorf("%w while seting next peer authentication time", setTimeErr)
+		}
+
+		setValidatorErr := sender.keysHolder.SetValidatorState(pkBytes, true)
+		if setValidatorErr != nil {
+			return fmt.Errorf("%w while seting validator state", setValidatorErr)
+		}
+	}
+
+	sender.sendData(index, pkBytes, data, isHardforkTriggered)
+
+	return nil
+}
+
+func (sender *multikeyPeerAuthenticationSender) processIfShouldSend(pkBytes []byte, currentTimeAsUnix int64) bool {
+	if !sender.keysHolder.IsKeyManagedByCurrentNode(pkBytes) {
+		return false
+	}
+	isValidatorNow, shardID := sender.getIsValidatorStatusAndShardID(pkBytes)
+	isHardforkSource := sender.isHardforkSource(pkBytes)
+	oldIsValidator, err := sender.keysHolder.IsKeyValidator(pkBytes)
+	if err != nil {
+		return false
+	}
+
+	err = sender.keysHolder.SetValidatorState(pkBytes, isValidatorNow)
+	if err != nil {
+		return false
+	}
+
+	if !isValidatorNow && !isHardforkSource {
+		return false
+	}
+	if shardID != sender.shardCoordinator.SelfId() {
+		return false
+	}
+
+	nextTimeToCheck, err := sender.keysHolder.GetNextPeerAuthenticationTime(pkBytes)
+	if err != nil {
+		return false
+	}
+
+	timeToCheck := nextTimeToCheck.Unix() < currentTimeAsUnix
+	if timeToCheck {
+		return true
+	}
+	if !oldIsValidator && isValidatorNow {
+		return true
+	}
+
+	return false
+}
+
+func (sender *multikeyPeerAuthenticationSender) prepareMessage(pkBytes []byte, privateKey crypto.PrivateKey) ([]byte, bool, error) {
+	p2pSkBytes, pid, err := sender.keysHolder.GetP2PIdentity(pkBytes)
+	if err != nil {
+		return nil, false, err
+	}
+
+	msg := &heartbeat.PeerAuthentication{
+		Pid:    pid.Bytes(),
+		Pubkey: pkBytes,
+	}
+
+	hardforkPayload, isTriggered := sender.getHardforkPayload()
```

Comment:
this function contains a lot of duplicated code with the original peerAuthenticationSender implementation. Can we extract duplicated code as seen in the heartbeat sender future implementation?